lib/synapse/config_generator/haproxy.rb

```diff
@@ -1055,9 +1072,28 @@ def generate_backend_stanza(watcher, config)
             @restart_required = true
           end
         end
+
         backends[backend_name] = backend.merge('enabled' => true)
+
+        # If the the registry defines the haproxy_server_id that must be preferred.
+        server_opts = (backend['haproxy_server_options'] || 'no match').split(' ')
+        @server_id_map[watcher.name][backend_name] = server_opts[server_opts.index('id') + 1].to_i if server_opts.include?('id')
+        # This is going to be an issue ... and haproxy will fail if this happens ... oh well
+        @server_id_map[watcher.name][backend_name] = backend['haproxy_server_id'] if backend['haproxy_server_id']
       end
 
+      # Now that we know the maximum possible existing haproxy_server_id for
+      # this backend, we can set any that don't exist yet.
+      watcher.backends.each do |backend|
+        backend_name = construct_name(backend)
+        @server_id_map[watcher.name][backend_name] ||= (
+          ((@server_id_map[watcher.name].values + [0]).compact.max) % @max_server_id + 1
```

Comment:
This isn't very efficient and can lead to a problem with three backends (a,b,c) but two set haproxy_server_options with id = 1 (for a) and the other id = max_server_id (for b). This loop could set id (for c) to also be 1 causing a conflict. This situation is nearly impossible but should still be addressed. A reverse index id_server_map[watcher.name][id] -> backend_name might fix this.